psi4/src/psi4/libfock/DFJCOSK.cc

```diff
@@ -0,0 +1,1012 @@
+ * the corresponding files.
+ *
+ * This file is part of Psi4.
+ *
+ * Psi4 is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, version 3.
+ *
+ * Psi4 is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License along
+ * with Psi4; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * @END LICENSE
+ */
+
+#include "jk.h"
+#include "psi4/libqt/qt.h"
+#include "psi4/libfock/cubature.h"
+#include "psi4/libfock/points.h"
+#include "psi4/libmints/vector.h"
+#include "psi4/libmints/basisset.h"
+#include "psi4/libmints/electrostatic.h"
+#include "psi4/libmints/mintshelper.h"
+#include "psi4/libmints/molecule.h"
+#include "psi4/libmints/integral.h"
+#include "psi4/liboptions/liboptions.h"
+#include "psi4/lib3index/dftensor.h"
+
+#include <vector>
+#include <map>
+#include <algorithm>
+#ifdef _OPENMP
+#include <omp.h>
+#endif
+
+using namespace psi;
+
+namespace psi {
+
+SharedMatrix compute_numeric_overlap(std::shared_ptr<DFTGrid> grid, std::shared_ptr<BasisSet> primary) {
+
+    // DOI 10.1063/1.3646921, EQ. 9
+
+    int nbf = primary->nbf();
+    BasisFunctions bf_computer(primary, grid->max_points(), grid->max_functions());
+    auto S_num = std::make_shared<Matrix>("Numerical Overlap", nbf, nbf);
+
+    // This loop could be parallelized over blocks of grid points. However, the cost of the loop is
+    // so small (< 10 seconds for a 200 heavy atom system), parallelism isn't necessary
+
+    for (size_t bi = 0; bi < grid->blocks().size(); bi++) {
+
+        // grid points in this block
+        auto block = grid->blocks()[bi];
+        int npoints_block = block->npoints();
+        int nbf_block = block->local_nbf();
+        auto w = block->w();
+
+        // compute basis functions at these grid points
+        bf_computer.compute_functions(block);
+        auto point_values = bf_computer.basis_values()["PHI"];
+
+        // resize the buffer of basis function values
+        Matrix X_block("phi_g,u", npoints_block, nbf_block);  // points x nbf_block
+        auto X_blockp = X_block.pointer();
+        for (size_t p = 0; p < npoints_block; p++) {
+            for (size_t k = 0; k < nbf_block; k++) {
+                X_blockp[p][k] = point_values->get(p, k) * std::sqrt(w[p]);
+            }
+        }
+
+        // significant basis functions at these grid points
+        std::vector<int> bf_map = block->functions_local_to_global();
+
+        auto S_num_block = linalg::doublet(X_block, X_block, true, false);
+        auto S_num_blockp = S_num_block.pointer();
```

Comment:
For my own edification, is it any faster to grab a pointer object and access by index rather than call `S_num_blockp->get`? If it is, why aren't we doing the same trick with `S_num[mu][nu] +=`?

Reply:
Any function call adds an overhead, iirc, but the compiler can inline the function to eliminate that.